packages/require/lib/typescript.ts

```diff
@@ -0,0 +1,137 @@
+import path from "path";
+
+import type TSNode from "ts-node";
+import TruffleConfig from "@truffle/config";
+import { readFileSync } from "fs";
+
+import originalRequire from "original-require";
+
+import { TsNodeDependencyError } from "./types";
+
+import Debug from "debug";
+
+const debug = Debug("@truffle/require:typescript");
+
+const _tsExtensionExpr = /^\.(cts|tsx?)$/i;
+
+let tsNode: typeof TSNode | null = null;
+try {
+  // We use `originalRequire` here to ensure that we use ts-node from the
+  // user's environment rather than resolving our own (should we
+  // accidentally bundle ts-node)
+  tsNode = originalRequire("ts-node");
+} catch {}
+
+/**
+ * @param conf the instance of `@truffle/config` for the user's project
+ *
+ * @param sourceFilePath the path to the source module to be compiled
+ *
+ * @param context the global context in which the compiled script will run
+ *
+ * @returns the TypeScript code stored in the file at location `sourceFilePath`
+ *   transpiled to JavaScript when `scriptPath` has a TypeScript file extension,
+ *   otherwise it returns the contents of the `scriptPath` file.
+ */
+export function compile(conf: TruffleConfig, sourceFilePath: string): string {
+  const source = readFileSync(sourceFilePath, { encoding: "utf-8" });
+
+  // we only compile TS files, so just return the source unless we are dealing
+  // with one of those
+  if (!_tsExtensionExpr.test(path.extname(sourceFilePath))) {
+    debug(`${sourceFilePath} is not a TS file, returning unmodified source`);
+    return source;
+  }
+
+  const compilationService = _getOrCreateCompilationService(
+    conf,
+    sourceFilePath
+  );
+
+  debug(`compiling ${sourceFilePath}`);
+  return compilationService.compile(source, sourceFilePath);
+}
+
+/**
+ * Creates and caches an instance of TypeScript compiler service that has been
+ * initialized with the global context for the module that we'll be executing.
+ *
+ * @param conf the instance of `@truffle/config` for the user's project
+ *
+ * @param sourceFilePath the path to the source module being compiled
+ *
+ * @param context the global context in which the compiled script will run
+ *
+ * @returns an instance of a TypeScript compiler service (TS Compiler API)
+ */
+function _getOrCreateCompilationService(
+  conf: TruffleConfig,
+  sourceFilePath: string
+): TSNode.Service {
+  if (!tsNode) {
+    // TODO: add error chaining support
```

Comment:
What does error chaining support entail? Maybe this can become an issue instead of a TODO so it's not lost in the codebase, and we get eyes on it.

Reply:
#5335 5335